smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/IncomingKafkaBatchRecords.java

```diff
@@ -0,0 +1,106 @@
+package io.smallrye.reactive.messaging.kafka;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.common.TopicPartition;
+import org.eclipse.microprofile.reactive.messaging.Metadata;
+
+import io.smallrye.mutiny.Uni;
+import io.smallrye.reactive.messaging.kafka.api.IncomingKafkaBatchRecordsMetadata;
+import io.smallrye.reactive.messaging.kafka.commit.KafkaCommitHandler;
+import io.smallrye.reactive.messaging.kafka.fault.KafkaFailureHandler;
+
+public class IncomingKafkaBatchRecords<K, T> implements KafkaBatchRecords<K, T> {
+
+    private final Metadata metadata;
+    private final List<KafkaRecord<K, T>> incomingRecords;
+    private final Map<TopicPartition, KafkaRecord<K, T>> latestOffsetRecords;
+
+    public IncomingKafkaBatchRecords(ConsumerRecords<K, T> records, KafkaCommitHandler commitHandler,
+            KafkaFailureHandler onNack, boolean cloudEventEnabled, boolean tracingEnabled) {
+        List<IncomingKafkaRecord<K, T>> incomingRecords = new ArrayList<>();
+        Map<TopicPartition, IncomingKafkaRecord<K, T>> latestOffsetRecords = new HashMap<>();
+        for (TopicPartition partition : records.partitions()) {
+            IncomingKafkaRecord<K, T> latestOffsetRecord = null;
+            for (ConsumerRecord<K, T> record : records.records(partition)) {
+                IncomingKafkaRecord<K, T> rec = new IncomingKafkaRecord<>(record, commitHandler, onNack,
+                        cloudEventEnabled, tracingEnabled);
+                incomingRecords.add(rec);
+                latestOffsetRecord = latestOffsetRecords.compute(partition,
```

Comment:
Just out of curiosity, wouldn't `latestOffsetRecords.put(partition, rec)` work as well? Records are consumed in sequence, so unconditionally putting each record into the map ends up with having the last record there, for each partition?

Reply:
(And of course unconditionally setting `latestOffsetRecord = rec`.)

Reply:
Yes, I may have over-thought this, or not enough :) 
Not having an ascending order inside a partition of `ConsumerRecords` would be very problematic.